src/runtime/task/mod.rs

```diff
@@ -248,27 +264,45 @@ impl Task {
         self.local_storage.pop()
     }
 
-    /// Park the task if its park token is unavailable. Returns true if the token was unavailable.
+    /// Park the task if its park token is unavailable. If the task blocks, then it will be woken up
+    /// when the token becomes available or spuriously without consuming the token (see the
+    /// documentation for [`std::thread::park`], which says that "it may also return spuriously,
+    /// without consuming the token"). Returns true if the execution should switch to a different
+    /// task (e.g., if the token was unavailable).
     pub(crate) fn park(&mut self) -> bool {
-        match self.park_state {
-            ParkState::Unparked => {
-                self.park_state = ParkState::Unavailable;
-                false
-            }
-            ParkState::Unavailable => {
-                self.park_state = ParkState::Parked;
-                self.block();
-                true
-            }
-            ParkState::Parked => unreachable!("cannot park a task that's already parked"),
+        assert!(
+            !self.park_state.blocked_in_park,
+            "task cannot park while already parked"
+        );
+        assert!(!self.blocked(), "task cannot park while blocked by something else");
+
+        if self.park_state.token_available {
+            self.park_state.token_available = false;
+            false
+        } else {
+            self.park_state.blocked_in_park = true;
+            self.block(true);
```

Comment:
Should this also set `self.has_yielded`? Otherwise a PCT scheduler will test spurious wakeups (almost) every time, because PCT (almost) always schedules the same thread if it's still runnable and not yielding.

Reply:
Er, `has_yielded` is in `ExecutionState`, but you get the idea.

Reply:
Yep, I added a call to `ExecutionState::request_yield()` in the `park()` handler where the `ExecutionState` is available.